items/src/main/scala/weco/api/items/services/ItemUpdateService.scala

```diff
@@ -1,62 +1,91 @@
 package weco.api.items.services
 
-import grizzled.slf4j.Logging
-import weco.api.stacks.services.SierraService
 import weco.catalogue.internal_model.identifiers.{
   IdState,
   IdentifierType,
   SourceIdentifier
 }
-import weco.catalogue.internal_model.locations.{
-  AccessCondition,
-  PhysicalLocation
-}
 import weco.catalogue.internal_model.work.{Item, Work, WorkState}
 
 import scala.concurrent.{ExecutionContext, Future}
 
+trait ItemUpdater {
+  val identifierType: IdentifierType
+  def updateItems(
+    items: Seq[Item[IdState.Identified]]
+  ): Future[Seq[Item[IdState.Identified]]]
+}
+
 class ItemUpdateService(
-  sierraService: SierraService
-)(implicit executionContext: ExecutionContext)
-    extends Logging {
-
-  private def isSierraId(sourceIdentifier: SourceIdentifier) =
-    sourceIdentifier.identifierType == IdentifierType.SierraSystemNumber
-
-  private def updateAccessCondition(
-    item: Item[IdState.Minted],
-    accessCondition: AccessCondition
-  ) =
-    item.locations.map {
-      case physicalLocation: PhysicalLocation =>
-        physicalLocation.copy(
-          accessConditions = List(accessCondition)
-        )
-      case location => location
-    }
+  itemUpdaters: List[ItemUpdater]
+)(implicit executionContext: ExecutionContext) {
+
+  private final val itemUpdatesMap = itemUpdaters
+    .map(updater => updater.identifierType -> updater)
+    .toMap
+
+  private def buildItemsMap(
+    items: Seq[Item[IdState.Identified]]
+  ): Map[SourceIdentifier, Item[IdState.Identified]] =
+    items.map {
+      case item @ Item(IdState.Identified(_, srcId, _), _, _, _) =>
+        srcId -> item
+    } toMap
+
+  private def preserveOrder(
+    itemsWithIndex: Seq[(Item[IdState.Identified], Int)],
+    f: Seq[Item[IdState.Identified]] => Future[Seq[Item[IdState.Identified]]]
+  ) = {
+    val itemsMap = buildItemsMap(itemsWithIndex.map(_._1))
+    val indexes = itemsWithIndex.map(_._2)
 
-  private def refreshItem(srcId: SourceIdentifier, item: Item[IdState.Minted]) =
-    sierraService
-      .getAccessCondition(srcId)
-      .map {
-        case Right(Some(accessCondition)) =>
-          item.copy(
-            locations = updateAccessCondition(item, accessCondition)
+    val indexLookup = itemsMap.keys.zip(indexes).toMap
+    val items = itemsMap.values.toSeq
+
+    f(items).flatMap { updatedItems =>
+      val updatedItemsMap = buildItemsMap(updatedItems)
+      val updatedItemsWithIndex = indexLookup.map {
+        case (srcId, index) =>
+          updatedItemsMap.get(srcId).map { updatedItem =>
+            (updatedItem, index)
+          }
+      } flatten
```

Comment:
I think I know what's going on here and it's correct, but it's still a bit gnarly. Possibly I'm tired, but a few comments or type signatures would help me read this.

Reply:
Completely fair. I think this function and `updateItems` could do with some comments. There might be a little more work to make the code easier to read, but what's going on is compact and complicated, warranting a bit of written explanation.